src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperObject.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+public abstract class TagHelperObject
+{
+    private int _flags;
+
+    private protected const int ContainsDiagnosticsBit = 1 << 0;
+    private protected const int CaseSensitiveBit = 1 << 1;
+    private protected const int LastFlagBit = CaseSensitiveBit;
+
+    private protected bool HasFlag(int flag) => (_flags & flag) != 0;
+    private protected void SetFlag(int toSet) => ThreadSafeFlagOperations.Set(ref _flags, toSet);
+    private protected void ClearFlag(int toClear) => ThreadSafeFlagOperations.Clear(ref _flags, toClear);
+    private protected void SetOrClearFlag(int toChange, bool value) => ThreadSafeFlagOperations.SetOrClear(ref _flags, toChange, value);
```

Comment:
Looking through this PR, I don't see anywhere that needs the `Clear` functionality. Unless I missed some reference, the closest that anything comes to it is `TagHelperDescriptorExtensions.IsChildContentTagHelper` and `TagHelperDescriptorExtensions.IsComponentFullyQualifiedNameMatch`, both of which always set a non-null value and only use the nullity of their respective properties as an indication of whether it's been set. Given our move towards making these objects immutable after construction, I'd prefer if we remove the `Clear` functionality, and update those two extensions to behave correctly in the face of that change. It somewhat seems like we could just inline the functionality in the extension directly into the properties themselves.

Reply:
Great suggestion!

Reply:
Now that I have the flag clearing removed, the whole flag thing just seems super contrived. I'm going to rewrite this bit to be fully immutable and just compute the values up front in the constructor. We already do that for other flags and computing these should be cheap.